core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -535,7 +536,7 @@ class GroupMetadataManager(val brokerId: Int,
 
     val numExpiredOffsetsRemoved = inWriteLock(offsetExpireLock) {
       val expiredOffsets = offsetsCache.filter { case (groupTopicPartition, offsetAndMetadata) =>
-        offsetAndMetadata.expireTimestamp < startMs
+        offsetAndMetadata.expireTimestamp < startMs || !metadataCache.contains(groupTopicPartition.topicPartition.topic)
```

Comment:
Yes, definitely good to avoid holding multiple locks at the same time (even though this is a read lock, it still blocks if the write lock is held). Also, the current code acquires the read lock once per `GroupTopicPartition` in the `offsetsCache` versus acquiring the lock once for @hachikuji's proposed solution.